### ISSUE TITLE
CEPH warnings break cephRBDVolumeListSnapshots

### ISSUE THREAD

**Issue by amcduffee** [lxc/lxd#5651]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 18.04.2 LTS
 * The output of "lxc info" or if that fails:
```
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE-----
  certificate_fingerprint: ...
  driver: lxc
  driver_version: 3.1.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "false"
    shiftfs: "false"
    uevent_injection: "false"
    unpriv_fscaps: "true"
  kernel_version: 4.15.0-47-generic
  project: default
  server: lxd
  server_clustered: false
  server_pid: 82657
  server_version: "3.12"
  storage: ""
  storage_version: ""
```

# Issue description

It doesn't appear that the LXD snap is able to correctly lookup DNS SRV service records in order to determine the ceph monitor hosts on machines that use an empty ceph.conf.

# Steps to reproduce

 1. Create an appropriate DNS SRV record for the ceph monitor, as per:
   http://docs.ceph.com/docs/nautilus/rados/configuration/mon-lookup-dns/
 2. Verify that the LXD snap host can successfully query the service record:
```
$ systemd-resolve --service _ceph-mon._tcp.local
_ceph-mon._tcp.local: cephmon.local:6789 [priority=10, weight=60]
                                 10.0.1.1
                                 (_ceph-mon._tcp/local)

-- Information acquired via protocol DNS in 1.7ms.
-- Data is authenticated: no
```
 3. Make sure ceph.conf is empty so the ceph commands attempt to do DNS SRV lookup:
```
$ cat /etc/ceph/ceph.conf 
[global]
```
4. Check that ceph status works with an empty ceph.conf on the LXD snap host:
```
$ ceph -s
  cluster:
    id:     233a8738-8518-4229-8028-dea1718dfcdb
    health: HEALTH_OK
```
5. Try to add a ceph storage pool to the LXD snap and it will fail:
```
$ lxc storage create ceph ceph
Error: Failed to run: ceph --name client.admin --cluster ceph osd pool create ceph 32: no monitors specified to connect to.
Error connecting to cluster: ObjectNotFound
```

6. Explicitly set the ceph monitor host in ceph.conf on the LXD snap host:
```
$ cat /etc/ceph/ceph.conf 
[global]
mon host = 10.0.1.1
```
7. Try to add the ceph storage pool again and this time it will succeed:
```
$ lxc storage create ceph ceph
Storage pool ceph created
```

**Comment by stgraber** [comment]
That's normal, the built-in CEPH client is at version 10.1 which predates version 11 that introduces support for DNS SRV records. This will change once we move the entire LXD snap to an Ubuntu 18.04 based rather than 16.04 as that will bump the CEPH client and libraries to 12.2.

**Comment by amcduffee** [comment]
Ahh, that makes sense.  Thanks for the clarification.

Any idea how long before that transition will occur?  I seem to be at a bit of an impasse here in terms of considering a new LXD & ceph setup.  I don't want to use the LXD snap because I have DNS SRV working to avoid manually administration of ceph.conf files.  Yet, if I stick with the native LXD deb package installed on 18.04.2 LTS the ceph storage pool has a lot of issue related to lxd copy & delete operations:

```
# lxc version
Client version: 3.0.3
Server version: 3.0.3
# lxc launch ubuntu:18.04 u1
Creating u1
Starting u1                                 
# lxc list
+------+---------+------+------+------------+-----------+------------+
| NAME |  STATE  | IPV4 | IPV6 |    TYPE    | SNAPSHOTS |  LOCATION  |
+------+---------+------+------+------------+-----------+------------+
| u1   | RUNNING |      |      | PERSISTENT | 0         |   host03   |
+------+---------+------+------+------------+-----------+------------+
# lxc copy u1 u2
Error: Unable to perform container live migration. CRIU isn't installed on the source server
# lxc stop u1
# lxc copy u1 u2
Error: Failed container creation: Error transferring container data: invalid character '-' after top-level value
# lxc list
+------+---------+------+------+------------+-----------+------------+
| NAME |  STATE  | IPV4 | IPV6 |    TYPE    | SNAPSHOTS |  LOCATION  |
+------+---------+------+------+------------+-----------+------------+
| u1   | STOPPED |      |      | PERSISTENT | 0         |   host03   |
+------+---------+------+------+------------+-----------+------------+
| u2   | STOPPED |      |      | PERSISTENT | 0         |   host03   |
+------+---------+------+------+------------+-----------+------------+
# lxc start u2
Error: Failed to run: /usr/lib/lxd/lxd forkstart u2 /var/lib/lxd/containers /var/log/lxd/u2/lxc.conf: 
Try `lxc info --show-log u2` for more info
# lxc delete u2
Error: Failed to delete RBD storage volume for container "u2" on storage pool "remote"
# lxc list
+------+---------+------+------+------------+-----------+------------+
| NAME |  STATE  | IPV4 | IPV6 |    TYPE    | SNAPSHOTS |  LOCATION  |
+------+---------+------+------+------------+-----------+------------+
| u1   | STOPPED |      |      | PERSISTENT | 0         |   host03   |
+------+---------+------+------+------------+-----------+------------+
| u2   | STOPPED |      |      | PERSISTENT | 0         |   host03   |
+------+---------+------+------+------------+-----------+------------+
```

**Comment by amcduffee** [comment]
So I think I tracked down the issue.  I had inadvertently introduced some centralized cluster configuration keys into the ceph mon config database that are immutable once they are created.  I couldn't find any way to get them purged and unfortunately they generate warning messages on stderr whenever 'rbd' commands are run.

The call stack for the 'Error: Failed to delete RBD storage volume for container "u2" on storage pool "remote"' error is:
```
MigrationSource()
cephRBDVolumeListSnapshots()
shared.RunCommand("rbd", ...)
exec.Command(name, arg...).CombinedOutput()
```
The CombinedOutput() is pulling in the warnings from stderr and breaking the json.Unmarshal() call in cephRBDVolumeListSnapshots().

Since I was working on an experimental setup I was able to do a fresh ceph deployment and the problems in my last comment went away.

I think this is a bug in LXD because cephRBDVolumeListSnapshots() shouldn't be trying to parse stderr as JSON since it is likely to have warning message.
